File: src/commands/AddIfMaxCommand.java
package commands;

import client.Application;
import client.OrganizationBuilder;
import commands.core.ServerAndClientSideCommand;
import lib.ExecutionStatus;
import lib.functions.CallbackFunction;
import organization.Organization;
import organization.OrganizationManager;

import java.util.Collections;
import java.util.Comparator;

public class AddIfMaxCommand extends ServerAndClientSideCommand {

    public AddIfMaxCommand(CallbackFunction callbackFunction, Application application, OrganizationManager organizationManager) {
        super(callbackFunction, application, organizationManager);
    }

    @Override
    protected void executeImplementation(String[] args, CallbackFunction callback) throws Exception {
        Organization newOrganization = OrganizationBuilder.constructOrganization(application.getBufferedReaderWithQueueOfStreams(), false);
        Organization maxOrganization = Collections.max(organizationManager.getOrganizations(),
                Comparator.comparing(Organization::getFullName));
        ExecutionStatus executionStatus = ExecutionStatus.FAILURE;

        if (maxOrganization.getFullName().compareTo(newOrganization.getFullName()) < 0) {
            organizationManager.add(newOrganization);
            executionStatus = ExecutionStatus.SUCCESS;
        }

        callback.invoke(executionStatus, null);
    }
}

File: src/commands/client_side/AddIfMaxCommand.kt
package commands.client_side;

import application.Application;
import application.OrganizationBuilder;
import commands.client_side.core.ServerAndClientSideCommand;
import commands.client_side.core.ClientCallbackFunction;
import OrganizationDatabase.Organization;
import OrganizationDatabase.OrganizationManagerInterface;

public class AddIfMaxCommand extends ServerAndClientSideCommand {

    public AddIfMaxCommand(ClientCallbackFunction clientCallbackFunction, Application application, OrganizationManagerInterface organizationDatabase) {
        super(clientCallbackFunction, application, organizationDatabase);
    }

    @Override
    protected void executeImplementation(String[] args, ClientCallbackFunction callback) throws Exception {
        Organization newOrganization = OrganizationBuilder.constructOrganization(
                application.getBufferedReaderWithQueueOfStreams(),
                false
        );

        callback.invoke(organizationDatabase.addIfMax(newOrganization), null);
    }
}
